### ISSUE TITLE
False positive and false negative in unnecessary_parentheses

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/linter#5112]
I created one issue instead of true because these might be related. It almost looks like a condition is reversed.

I believe the lint should produce a diagnostic for the following code, but does not.
```dart
class C {
  void m() {
    (1 + 2);
  }
}
```
removing the parentheses does not change the semantics of the code.

On the other hand, the lint should not produce a diagnostic for the following code, but does.
```dart
class C {
  void m() {
    ({1, 2, 3});
  }
}
```
removing the parentheses here causes the set literal to be interpreted as a block.

**Comment by mmcdon20** [comment]
I found another case relating to `unnecessary_parentheses` that I think might be a false negative.

If you have an immediately invoked function expression (IIFE) you can add an unnecessary set of parentheses and not get a lint warning.

```dart
(() {})();
```
